014-single-codebase.md

```diff
@@ -0,0 +1,68 @@
+
+- **Technical overhead:** Logic that straddles multiple domains is also inevitably more complicated, requiring network requests for what could otherwise be a function call or `JOIN` statement. It makes traditionally simple logic into distributed systems problems where we have to decide whether we'll trade consistency for performance or vice versa.
+
+- **Testing limitations:** We know we get the best bang for our buck from [high-level integration tests](https://github.com/DoSomething/rfcs/blob/master/004-testing.md), but separating into services forces us to use mocking extensively because the majority of our application (e.g. users in Rogue, or users & posts in Phoenix) is always an external dependency in any given test! This places a large burden on our intra-service contracts.
+
+- **Knowledge silos:** Finally, this separation reinforces silos (e.g. "you have to talk to Dave about that app" or "that's X application's responsibility, not ours"). We have a smaller team today than when we first started splitting out into a service oriented architecture, and it's important that we can all understand & comfortably work on the whole system.
+
+
+
+## Proposal
+
+We should consolidate our backend services (Northstar, Rogue, and Chompy) into **a single monolithic Laravel application.**
+
+- This will **simplify architectural decisions** & enable us to iterate faster. Instead of sorting new functionality into strict bounded contexts up front, we'll be able to simply place new entities into a single backend context. Architectural decisions will be less binding, as it's easier to refactor within a single codebase than across network boundaries.
+
+- It will also **simplify adding behavior that straddles contexts**, like clubs, by replacing network requests with function calls. We'll also be able to ensure better consistency by using more transactional validation.
+
+- Keeping more of our business entities in the same domain allows **easier integration testing**, since we can seed (and otherwise directly reference) data that would currently live across API boundaries, like user IDs & clubs.
+
+- We have less applications to keep up-to-date with security & framework upgrades!
+
+For services that may not make sense to fold into the monolith, like [GraphQL](https://github.com/DoSomething/graphql/) or [Gambit](https://github.com/DoSomething/gambit/), we should still collaborate on them [**as part of a single monorepo**](https://docs.google.com/document/d/1a3e0Wpv9qeRscspdbN4wVR4Chz5oJFb9HH69x2pngSg/edit), side-by-side with our consolidated backend application. This allows us to retain the technical benefits we've gotten from separating out these services, while collaborating in one place.
```

Comment:
Can you give me more context on what this means? I read through the monorepo doc but I'm not 100% sure I understand how they would continue to be separate services within the same repo.

Reply:
+1 I have the same question! I've never really felt that I understood the benefit of the monorepo idea and I'm curious if this would be part of this unified backend work or separate?

Reply:
From my perspective, the main benefits we've gotten from splitting out services are allowing them to use a different stack (e.g. using Node.js for Gambit & GraphQL) or scale independently (e.g. how we scale up Northstar and Rogue during SMS broadcasts, but we don't scale up Phoenix).

I _don't_ feel like the "separation of concerns" we get from putting them in separate GitHub repositories is nearly as useful though (and is perhaps even detrimental, since it means you have to make related changes in two or more separate pull requests).

So in cases where we'd want to use a different language or scaling strategy for some part of our stack, I'd still recommend keeping those apps in the same codebase (alongside things that may deploy elsewhere), even if we continue to deploy them separately. That way, for example, a change to GraphQL (a Node.js Lambda app) can happen right alongside the relevant changes to our Laravel backend APIs!